### ISSUE TITLE
Query cache issue when querying optional nodes and toString()

### ISSUE THREAD

**Issue by Brakebein** [neo4j/neo4j#12398]
**Neo4j Version:** 4.0.0  
**Operating System:** Windows 10 / Windows Server 2012 R2  
**API:** neo4j-driver 4.0.1 (nodejs)

I am facing an issue that after the Neo4j instance has been running for a while, a special cypher query fails to execute. After some testing and considering the error log, I think I could narrow it down to these lines of the cypher query:

```sql
// ...
OPTIONAL MATCH (e65)-[:P4]->(:E52)-[:P82]->(date0:E61)
// ...
WITH ... , date0, ...
// ...
RETURN
  // ...
  date0 {.*, from: toString(date0.from), to: toString(date0.to)} AS date
  // ...
```

`date0` represents a timespan and has (among others) the properties `from` and `to`, which are of type `date()`. For the return output I want them to be a string formatted `YYYY-MM-DD`, that's why I am using `toSting()`. Since `date0` is optional, it can sometimes be `NULL`. In this case, the query returns `date: null`.

This has always worked fine with Neo4j 3.5.x and it also works fine with Neo4j 4.0.0, at least for some time. With Neo4j 4.0.0, after a while, this query fails when the queried dataset includes nodes that are not connected with `date0:E61` node (i.e. `date0` is NULL). Then, it throws the following error, which leads me to the `toString()` function:

```
Neo.ClientError.Statement.TypeError
Neo4jError: Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE
```

I have also found out that after clearing the query cache, the query will work fine again for some time. This can be actually a few days, the total number of queries is quite low (as this is only a prototype/showcase application).

```sql
CALL db.clearQueryCaches()
```

This leads me to the assumption that there might be a bug with the query cache.

### Steps to reproduce
I am not sure, if this can be easily reproduced, since this error occurs after the Neo4j instance has been running for a while.

### Expected behavior
If `date0` is `NULL`, `date` should return `null`. No errors. Should work all the time, as it did with Neo4j 3.5.x.

### Actual behavior
After some time, query fails due to TypeError. Query cache needs to be cleared, so query will work again.

#### Error output/stacktrace from neo4j-driver:
```
Neo.ClientError.Statement.TypeError

 Neo4jError: Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE

    at captureStacktrace (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\result.js:263:15)
    at new Result (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\result.js:68:19)
    at newCompletedResult (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\transaction.js:437:10)
    at Object.run (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\transaction.js:274:14)
    at Transaction.run (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\transaction.js:117:32)
    at C:\ServerTools\UH4D-Server\modules\neo4j-request.js:32:59
    at TransactionExecutor._safeExecuteTransactionWork (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\internal\transaction-executor.js:134:22)
    at TransactionExecutor._executeTransactionInsidePromise (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\internal\transaction-executor.js:122:32)
    at C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\internal\transaction-executor.js:61:15
    at new Promise (<anonymous>)
    at TransactionExecutor.execute (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\internal\transaction-executor.js:60:14)
    at Session._runTransaction (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\session.js:292:40)
    at Session.readTransaction (C:\ServerTools\UH4D-Server\node_modules\neo4j-driver\lib\session.js:264:19)
    at Object.readTransaction (C:\ServerTools\UH4D-Server\modules\neo4j-request.js:32:32)
    at query (C:\ServerTools\UH4D-Server\routes\image.js:128:9)
    at Layer.handle [as handle_request] (C:\ServerTools\UH4D-Server\node_modules\express\lib\router\layer.js:95:5)
    at next (C:\ServerTools\UH4D-Server\node_modules\express\lib\router\route.js:137:13)
    at Route.dispatch (C:\ServerTools\UH4D-Server\node_modules\express\lib\router\route.js:112:3)
    at Layer.handle [as handle_request] (C:\ServerTools\UH4D-Server\node_modules\express\lib\router\layer.js:95:5)
    at C:\ServerTools\UH4D-Server\node_modules\express\lib\router\index.js:281:22
    at Function.process_params (C:\ServerTools\UH4D-Server\node_modules\express\lib\router\index.js:335:12)
    at next (C:\ServerTools\UH4D-Server\node_modules\express\lib\router\index.js:275:10)
```

#### Neo4j `query.log`:
```
2020-02-13 12:47:15.985+0000 ERROR id:1014 - 1780 ms: bolt-session	bolt	neo4j-javascript/0.0.0-dev		client/127.0.0.1:51289	server/127.0.0.1:7687>	uh4d - neo4j - 
			MATCH (image:E38:UH4D)
				OPTIONAL MATCH (image)<-[:P94]-(e65:E65)-[:P4]->(:E52)-[:P82]->(date:E61)
				WITH image, date
				WHERE date IS NULL OR date.to > date($from) AND date.from < date($to)
			WITH image
			MATCH (image)-[:P106]->(file:D9),
				(image)-[:P102]->(title:E35),
				(image)-[:P48]->(identifier:E42),
				(image)<-[:P94]-(e65:E65)
			OPTIONAL MATCH (e65)-[:P14]->(:E21)-[:P131]->(author:E82)
			OPTIONAL MATCH (e65)-[:P4]->(:E52)-[:P82]->(date0:E61)
			OPTIONAL MATCH (image)-[:P105]->(:E40)-[:P131]->(owner:E82)
			OPTIONAL MATCH (image)-[:P3]->(desc:E62)-[:P3_1]->(:E55 {id: "image_description"})
			OPTIONAL MATCH (image)-[:P3]->(misc:E62)-[:P3_1]->(:E55 {id: "image_miscellaneous"})
			OPTIONAL MATCH (image)-[:has_spatial]->(spatial:Spatial)
			OPTIONAL MATCH (image)-[:has_tag]->(tag:TAG)
			WITH image, file, title, identifier, author, date0, owner, desc, misc, spatial, collect(tag.id) AS tags WHERE (NOT image.id =~ ".*_dummy") RETURN image.id AS id,
				file,
				title.value AS title,
				identifier.permalink AS permalink,
				identifier.slub_cap_no AS captureNumber,
				author.value AS author,
				date0 {.*, from: toString(date0.from), to: toString(date0.to)} AS date,
				owner.value AS owner,
				desc.value AS description,
				misc.value AS misc,
				spatial,
				tags - {capNo: ['dummy:hide'], regexDate: [], excludes: [], regexOwner: ['(?i).*dummy:hide.*'], regexTag: ['(?i).*dummy:hide.*'], regexTitle: ['(?i).*dummy:hide.*'], from: '1679-01-01', includes: [], to: '2020-02-13', regexAuthor: ['(?i).*dummy:hide.*']} - {} Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE
org.neo4j.exceptions.ParameterWrongTypeException: Expected a String, Number, Boolean, Temporal or Duration, got: NO_VALUE
	at org.neo4j.cypher.operations.CypherFunctions.toString(CypherFunctions.java:1130)
	at org.neo4j.codegen.Expression40/0x0000000101b1b040.project(Unknown Source)
	at org.neo4j.cypher.internal.runtime.slotted.expressions.CompileWrappingProjection.project(CompiledExpressionConverter.scala:160)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.ProjectOperator.operate(ProjectOperator.scala:41)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.OperatorTask.operateWithProfile(Operator.scala:251)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.OperatorTask.operateWithProfile$(Operator.scala:241)
	at org.neo4j.cypher.internal.runtime.pipelined.operators.ProjectOperator.operateWithProfile(ProjectOperator.scala:23)
	at org.neo4j.cypher.internal.runtime.pipelined.PipelineTask.executeOperators(PipelineTask.scala:65)
	at org.neo4j.cypher.internal.runtime.pipelined.PipelineTask.executeWorkUnit(PipelineTask.scala:48)
	at org.neo4j.cypher.internal.runtime.pipelined.Worker.executeTask(Worker.scala:128)
	at org.neo4j.cypher.internal.runtime.pipelined.Worker.workOnQuery(Worker.scala:84)
	at org.neo4j.cypher.internal.runtime.pipelined.execution.CallingThreadExecutingQuery.request(CallingThreadExecutingQuery.scala:30)
	at org.neo4j.cypher.internal.PipelinedRuntime$PipelinedRuntimeResult.request(PipelinedRuntime.scala:326)
	at org.neo4j.cypher.internal.result.StandardInternalExecutionResult.request(StandardInternalExecutionResult.scala:88)
	at org.neo4j.cypher.internal.result.ClosingExecutionResult.request(ClosingExecutionResult.scala:135)
	at org.neo4j.bolt.runtime.AbstractCypherAdapterStream.handleRecords(AbstractCypherAdapterStream.java:105)
	at org.neo4j.bolt.v3.messaging.ResultHandler.onPullRecords(ResultHandler.java:41)
	at org.neo4j.bolt.v4.messaging.PullResultConsumer.consume(PullResultConsumer.java:42)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State.consumeResult(TransactionStateMachine.java:511)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine$State$2.streamResult(TransactionStateMachine.java:355)
	at org.neo4j.bolt.runtime.statemachine.impl.TransactionStateMachine.streamResult(TransactionStateMachine.java:92)
	at org.neo4j.bolt.v4.runtime.InTransactionState.processStreamResultMessage(InTransactionState.java:73)
	at org.neo4j.bolt.v4.runtime.AbstractStreamingState.processUnsafe(AbstractStreamingState.java:49)
	at org.neo4j.bolt.v4.runtime.InTransactionState.processUnsafe(InTransactionState.java:60)
	at org.neo4j.bolt.v3.runtime.FailSafeBoltStateMachineState.process(FailSafeBoltStateMachineState.java:48)
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.nextState(AbstractBoltStateMachine.java:143)
	at org.neo4j.bolt.runtime.statemachine.impl.AbstractBoltStateMachine.process(AbstractBoltStateMachine.java:91)
	at org.neo4j.bolt.messaging.BoltRequestMessageReader.lambda$doRead$1(BoltRequestMessageReader.java:90)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.lambda$enqueue$0(DefaultBoltConnection.java:151)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatchInternal(DefaultBoltConnection.java:240)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:175)
	at org.neo4j.bolt.runtime.DefaultBoltConnection.processNextBatch(DefaultBoltConnection.java:165)
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.executeBatch(ExecutorBoltScheduler.java:212)
	at org.neo4j.bolt.runtime.scheduling.ExecutorBoltScheduler.lambda$scheduleBatchOrHandleError$2(ExecutorBoltScheduler.java:195)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by fabio-s-franco** [comment]
I think our issues are related and maybe you can confirm you have a cache issue, by applying the same workaround, which is to disable the query cache entirely:

https://github.com/neo4j/neo4j/issues/12399#issuecomment-586280196

After that the problem does not happen again.

**Comment by Brakebein** [comment]
I just did some additional tests and I can confirm that the problem disappears after disabling the query cache by setting `dbms.query_cache_size=0` in the config as @fabio-510 suggested in #12399 .

Our cypher queries are indeed very similar: we have both an `OPTIONAL MATCH` clause in combination with a map projection in `WITH` or `RETURN` section.

I also have counted the number of query executions until it starts throwing the error: I could run the query succesfully exact 10 times (just like @fabio-510 mentioned it, too).

**Comment by pontusmelke** [comment]
@Brakebein and @fabio-510 thanks a lot for reporting this. Your issues are similar but not exactly the same and I haven't verified we have fixed this issue yet.  Setting the `query_cache_size=0` will significantly reduce throughput since we will always replan which adds a big overhead to each execution of the query. As a temporary workaround you can prepend `CYPHER expressionEngine=INTERPRETED` to the query.

And yes you are exactly right that it happens after 10 times, there is a jitting mechanism that after 10 runs of the query will generate specialized byte code for the expression, in this case for

```
image.id AS id,
				file,
				title.value AS title,
				identifier.permalink AS permalink,
				identifier.slub_cap_no AS captureNumber,
				author.value AS author,
				date0 {.*, from: toString(date0.from), to: toString(date0.to)} AS date,
				owner.value AS owner,
				desc.value AS description,
				misc.value AS misc,
				spatial,
				tags - {capNo: ['dummy:hide'], regexDate: [], excludes: [], regexOwner: ['(?i).*dummy:hide.*'], regexTag: ['(?i).*dummy:hide.*'], regexTitle: ['(?i).*dummy:hide.*'], from: '1679-01-01', includes: [], to: '2020-02-13', regexAuthor: ['(?i).*dummy:hide.*']}
```

**Comment by Brakebein** [comment]
Has this issue been fixed in the `4.0.1` release? The changelog states:

> Fix issue that could lead to errors when using map projections with projections that evaluates to `null`.
> $fixes: 2398

https://github.com/neo4j/neo4j/wiki/Neo4j-4.0-changelog

The issue number should probably be 12398 and this issue can be closed?

**Comment by sherfert** [comment]
@Brakebein indeed!
